app/src/main/java/io/apicurio/registry/store/RegistryStorage.java

```diff
@@ -0,0 +1,78 @@
+ */
+
+package io.apicurio.registry.store;
+
+import java.util.List;
+
+import io.apicurio.registry.rest.beans.ArtifactMetaData;
+import io.apicurio.registry.rest.beans.EditableMetaData;
+import io.apicurio.registry.rest.beans.Rule;
+import io.apicurio.registry.rest.beans.VersionMetaData;
+
+/**
+ * The storage layer for the registry.
+ * @author eric.wittmann@gmail.com
+ */
+public interface RegistryStorage {
+
+    public ArtifactMetaData createArtifact(String artifactId, String content) throws ArtifactAlreadyExistsException, RegistryStorageException;
+    
+    public void deleteArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public StoredArtifact getArtifact(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public ArtifactMetaData updateArtifact(String artifactId, String content) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public ArtifactMetaData getArtifactMetaData(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public void updateArtifactMetaData(String artifactId, EditableMetaData metaData) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public List<Rule> getArtifactRules(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public void createArtifactRule(String artifactId, Rule rule) throws ArtifactNotFoundException, ArtifactRuleAlreadyExistsException, RegistryStorageException;
+    
+    public void deleteArtifactRules(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public Rule getArtifactRule(String artifactId, String ruleName) throws ArtifactNotFoundException, RuleNotFoundException, RegistryStorageException;
+    
+    public void deleteArtifactRule(String artifactId, String ruleNam) throws ArtifactNotFoundException, RuleNotFoundException, RegistryStorageException;
+    
+    public List<Long> getArtifactVersions(String artifactId) throws ArtifactNotFoundException, RegistryStorageException;
+    
+    public StoredArtifact getArtifactVersion(String artifactId, long version) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public void deleteArtifactVersion(String artifactId, long version) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public VersionMetaData getArtifactVersionMetaData(String artifactId, long version) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public void updateArtifactVersionMetaData(String artifactId, long version, EditableMetaData metaData) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
+    
+    public void deleteArtifactVersionMetaData(String artifactId, long version) throws ArtifactNotFoundException, VersionNotFoundException, RegistryStorageException;
```

Comment:
Do we need a method to delete metadata in addition to updating them? Would the `.../meta` endpoint return an error or `{}` if there are not metadata set?
In this case, the metadata are encoded as a single POJO/bean. The delete method would make more sense if the metadata are themselves a list of items, similar to the way rules work.

Reply:
I hadn't considered handling the meta-data as a collection... which I think is what you are suggesting here.  I could be convinced that is a better approach, although it would require multiple REST calls by the client in order to set the values.

Thoughts @alesj , @jsenko ?

Reply:
How or why would one update metadata?
Imho, I don't see the need for collection of metadatas.